Add support for custom paths and encrypted persistent Realms

swaffoja opened the pull request:
PR for issue/proposal #3 

`local_realm_provider.dart` contains examples for various configuration options:

- In-memory Realm
- Default persistent Realm
- Custom file name in default directory
- Subdirectory under default directory
- Custom file name and subdirectory
- Encrypted Realm

swaffoja committed:
Allow file directory and/or file name realm configuration

swaffoja committed:
Add example for file realm configuration

swaffoja committed:
Support iOS file name configuration

Stub

swaffoja committed:
Add support for custom directory paths and custom file names

swaffoja committed:
Implement encrypted realm iOS

swaffoja committed:
Add Android support for file directory, name, and encryption

swaffoja commented on example/lib/helpers/local_realm_provider.dart:
```diff
@@ -39,3 +45,88 @@ class _RealmProviderState extends State<RealmProvider> {
+  @override
+  void initState() {
+    super.initState();
+    final encryptionKey = _secureRandom(64);
```
Note that the example generates a new encryption key every time the example is run. You will get a fatal error if you run the example a 2nd time because the keys will not match. It might be better to write the 1st encryption key to a file for demo purposes so someone can re-open the encrypted Realm.

mogol replied:
Why don't use constant in demo?

swaffoja replied:
Great idea. Will update.

swaffoja commented on android/src/main/java/com/it_nomads/flutter_realm/FlutterRealm.java:
```diff
@@ -31,11 +32,25 @@
+        } else if (fileDirectory != null || fileName != null) {
+            if (fileDirectory != null) {
+                RealmConfiguration tmp = builder.build();
+                File file = new File(tmp.getRealmDirectory(), fileDirectory);
```
I'm not sure if `tmp.getRealmDirectory()` is the most efficient way to get the default Realm directory. I would prefer to call `builder.build()` only once...

mogol replied:
What is the case for custom directory?

swaffoja replied:
From Realm:

> The most common location to store writable Realm files is the “Documents” directory on iOS and the “Application Support” directory on macOS. Please respect Apple’s iOS Data Storage Guidelines, which recommend that if documents that can be regenerated by the app should be stored in the <Application_Home>/Library/Caches directory. If a custom URL is used to initialize a Realm, it must describe a location with write permissions.

> For example, suppose you have an application where users have to log in to your web backend, and you want to support quickly switching between accounts. You could give each account its own Realm file that will be used as the default Realm by doing the following...

I think two use cases for custom directories:

1. Your app needs to cache some non-user generated data and the cached data can be regenerated. The realm should technically be in the `.../Library/Caches` directory on iOS. Note that this PR does not support paths other than the standard documents directory on iOS or Android. If someone needs that flexibility in the future I'd suggest just adding a `cacheDirectory` property uses the appropriate base directory on iOS and Android.
1. My app allows multiple users on a single device which each user belonging to 1 or more projects. Data is sensitive and each user/project must have its own realm.
```
/user1/projectA/project.realm
/user1/projectA/additional_project_files...
/user1/projectB/project.realm
/user1/projectB/additional_project_files...
/user2/projectA/project.realm
etc...
```

mogol replied:
How about defining full path (including file name) in Dart code? 

It simplifies plugin (native, duplicated) code in iOS and Android and resolves problems what if I want to use TMP, DOC,CACHE or another folder.

swaffoja replied:
I can do that. I started out that way but then switched to the current approach. My thought was that 95% of the time a developer would want to write to the DOC directory. Current approach made it so you don't need to use [path_provider](https://pub.dev/packages/path_provider). Full path is more flexible though.

Side note, it may be awhile before I get around to making changes.

mogol commented:
@swaffoja thanks for PR!
I will check it soon, somehow I haven't seen notifications

mogol commented:
Could you add integration tests to your cases? 
It should help to keep the plugin workable, e.g. when Realm version is updated.

swaffoja reviewed (commented):
Yes. I'll add tests.

swaffoja committed:
Simplify file configuration

fileDirectory and fileName merged into filePath

swaffoja committed:
Add encrypted realm provider and storage

swaffoja committed:
Add integration test for encrypted realm

swaffoja committed:
Add unit test for encryption

swaffoja committed:
Update features to include encrypted realm

swaffoja commented:
- Reverted `local_realm_provider.dart` to original 
- Added `encrypted_realm_provider.dart` to illustrate encryption and full file path usage
- Added integration test for encrypted realm provider
- Added unit test for encryption

shinriyo commented:
@mogol are you maintaining the repository yet?

mogol commented:
As of today, I do not use it and do not update.

shinriyo commented:
@mogol Do you find other one? Or, official team released?

mogol commented:
I haven't heard anything, I did some experiments and prototypes, after that I stopped using Realm as it doesn't fit my requirements
